components/utils.py

```diff
@@ -1211,33 +1229,263 @@ def start_service(service_name, append_prefix=True, ignore_restart_fail=False):
+        for sub in cert.get('subject', ()):
+            for key, value in sub:
+                # XXX according to RFC 2818, the most specific Common Name
+                # must be used.
+                if key == 'commonName':
+                    if _dnsname_match(value, hostname):
+                        return
+                    dnsnames.append(value)
+
+    _no_hostname_match(hostname, dnsnames)
+
+
+class CertValidatingHTTPSConnection(httplib.HTTPConnection):
+    default_port = httplib.HTTPS_PORT
+
+    def __init__(self, host, port=None, ca_certs=None, **kwargs):
+        httplib.HTTPConnection.__init__(self, host, port, **kwargs)
+        self.ca_certs = ca_certs
+        if ca_certs is not None:
+            self.cert_reqs = ssl.CERT_REQUIRED
+        else:
+            self.cert_reqs = ssl.CERT_NONE
+
+    def connect(self):
+        httplib.HTTPConnection.connect(self)
+        if self.ca_certs:
+            self.sock = ssl.wrap_socket(self.sock,
+                                        ca_certs=self.ca_certs,
+                                        cert_reqs=self.cert_reqs)
+        else:
+            self.sock = ssl.wrap_socket(self.sock)
+
+        if self.cert_reqs & ssl.CERT_REQUIRED:
+            cert = self.sock.getpeercert()
+            hostname = self.host.split(':', 0)[0]
+            match_hostname(cert, hostname)
+
+
+class VerifiedHTTPSHandler(urllib2.HTTPSHandler):
+    def __init__(self, **kwargs):
+        urllib2.HTTPSHandler.__init__(self)
+        self._connection_args = kwargs
+
+    def https_open(self, req):
+        def https_class_wrapper(host, **kwargs):
+            full_kwargs = dict(self._connection_args)
+            full_kwargs.update(kwargs)
+            return CertValidatingHTTPSConnection(host, **full_kwargs)
+
+        return self.do_open(https_class_wrapper, req)
+
+
+def build_urlopen(cert_file=None):
+    if not cert_file:
+        return urllib2.urlopen
+
+    handler = VerifiedHTTPSHandler(ca_certs=cert_file)
+    opener = urllib2.build_opener(handler)
+    return opener.open
+
+
 def http_request(url, data=None, method='PUT',
-                 headers=None, timeout=None, should_fail=False):
+                 headers=None, timeout=None, should_fail=False,
+                 cert_file=None):
     headers = headers or {}
     request = urllib2.Request(url, data=data, headers=headers)
     request.get_method = lambda: method
+
+    urlopen = build_urlopen(cert_file)
+
     try:
         if timeout:
-            return urllib2.urlopen(request, timeout=timeout)
-        return urllib2.urlopen(request)
+            return urlopen(request, timeout=timeout)
+        return urlopen(request)
     except urllib2.URLError as e:
         if not should_fail:
             ctx.logger.error('Failed to {0} {1} (reason: {2})'.format(
                 method, url, e.reason))
 
 
+def rest_service_url():
+    manager_config = ctx_factory.get('manager-config')
+    security_config = manager_config['security']
```

Comment:
whats the purpose of using get in the previous line if this one assumes the value was indeed there? :)